README.md
# Becoming a teacher design history

https://bat-design-history.herokuapp.com/

[Design history](https://github.com/fofr/govuk-design-history) is an app built using the [GOV.UK prototyping kit](https://github.com/alphagov/govuk-prototype-kit).

## Document your designs

This repository makes it easy to:

* screenshot your designs
* create pages of screenshots to document designs
* document designs using the [GOV.UK Design System](https://design-system.service.gov.uk/)
* make designs shareable and linkable

## Take screenshots

### Configure

In [scripts/screenshot.js](scripts/screenshot.js) you should set:
* the domain you want to capture (eg a locally running prototype at localhost:3000 or an Heroku app)
* the paths on that domain with the titles you want to give them

Example paths:
```js
paths = [
  { title: 'Index page', path: '/'},
  { title: 'Terms and conditions', path: '/terms-conditions'}
]
```

### Run

```bash
node scripts/screenshot.js "a-generated-example"
```

This will:
* visit each page and save a screenshot
* save screenshots in a new directory
* generate an index page with screenshots listed in order

Example output:
```
node scripts/screenshot.js "a-generated-example"
Index generated: app/views/a-generated-example/index.html
https://design-history.herokuapp.com/
 >> app/assets/images/a-generated-example/design-history-index.png
```

[Example commit](/commit/eba8f4fbcb11e3f50b3084711ad9f90ebc59898e), [example page](https://design-history.herokuapp.com/a-generated-example).

### Document

When the index page and images are generated it’s easy to begin documenting those designs.

### Publish

You can easily publish your design history to Heroku to create a linkable, shareable design artefact.


CHANGELOG.md
# 0.0.1

- Create design history


server.js
// Core dependencies
const path = require('path')

// NPM dependencies
const bodyParser = require('body-parser')
const dotenv = require('dotenv')
const express = require('express')
const nunjucks = require('nunjucks')
const sessionInCookie = require('client-sessions')
const sessionInMemory = require('express-session')
const cookieParser = require('cookie-parser')
const marked = require('marked')

// Run before other code to make sure variables from .env are available
dotenv.config()

// Local dependencies
const middleware = [
  require('./lib/middleware/authentication/authentication.js'),
  require('./lib/middleware/extensions/extensions.js')
]
const config = require('./app/config.js')
const documentationRoutes = require('./docs/documentation_routes.js')
const packageJson = require('./package.json')
const routes = require('./app/routes.js')
const utils = require('./lib/utils.js')
const extensions = require('./lib/extensions/extensions.js')

// Variables for v6 backwards compatibility
// Set false by default, then turn on if we find /app/v6/routes.js
var useV6 = false
var v6App
var v6Routes

try {
  v6Routes = require('./app/v6/routes.js')
  useV6 = true
} catch (e) {
  // No routes.js in app/v6 so we can continue with useV6 false
}

const app = express()
const documentationApp = express()

if (useV6) {
  console.log('/app/v6/routes.js detected - using v6 compatibility mode')
  v6App = express()
}

// Set cookies for use in cookie banner.
app.use(cookieParser())
documentationApp.use(cookieParser())
const handleCookies = utils.handleCookies(app)
app.use(handleCookies)
documentationApp.use(handleCookies)

// Set up configuration variables
var releaseVersion = packageJson.version
var env = (process.env.NODE_ENV || 'development').toLowerCase()
var useAutoStoreData = process.env.USE_AUTO_STORE_DATA || config.useAutoStoreData
var useCookieSessionStore = process.env.USE_COOKIE_SESSION_STORE || config.useCookieSessionStore
var useHttps = process.env.USE_HTTPS || config.useHttps
var gtmId = process.env.GOOGLE_TAG_MANAGER_TRACKING_ID

useHttps = useHttps.toLowerCase()

var useDocumentation = (config.useDocumentation === 'true')

// Promo mode redirects the root to /docs - so our landing page is docs when published on heroku
var promoMode = process.env.PROMO_MODE || 'false'
promoMode = promoMode.toLowerCase()

// Disable promo mode if docs aren't enabled
if (!useDocumentation) promoMode = 'false'

// Force HTTPS on production. Do this before using basicAuth to avoid
// asking for username/password twice (for `http`, then `https`).
var isSecure = (env === 'production' && useHttps === 'true')
if (isSecure) {
  app.use(utils.forceHttps)
  app.set('trust proxy', 1) // needed for secure cookies on heroku
}

middleware.forEach(func => app.use(func))

// Set up App
var appViews = extensions.getAppViews([
  path.join(__dirname, '/app/views/'),
  path.join(__dirname, '/lib/')
])

var nunjucksConfig = {
  autoescape: true,
  noCache: true,
  watch: false // We are now setting this to `false` (it's by default false anyway) as having it set to `true` for production was making the tests hang
}

if (env === 'development') {
  nunjucksConfig.watch = true
}

nunjucksConfig.express = app

var nunjucksAppEnv = nunjucks.configure(appViews, nunjucksConfig)

// Add Nunjucks filters
utils.addNunjucksFilters(nunjucksAppEnv)

// Set views engine
app.set('view engine', 'html')

// Middleware to serve static assets
app.use('/public', express.static(path.join(__dirname, '/public')))

// Serve govuk-frontend in from node_modules (so not to break pre-extenstions prototype kits)
app.use('/node_modules/govuk-frontend', express.static(path.join(__dirname, '/node_modules/govuk-frontend')))

// Set up documentation app
if (useDocumentation) {
  var documentationViews = [
    path.join(__dirname, '/node_modules/govuk-frontend/'),
    path.join(__dirname, '/node_modules/govuk-frontend/components'),
    path.join(__dirname, '/docs/views/'),
    path.join(__dirname, '/lib/')
  ]

  nunjucksConfig.express = documentationApp
  var nunjucksDocumentationEnv = nunjucks.configure(documentationViews, nunjucksConfig)
  // Nunjucks filters
  utils.addNunjucksFilters(nunjucksDocumentationEnv)

  // Set views engine
  documentationApp.set('view engine', 'html')
}

// Support for parsing data in POSTs
app.use(bodyParser.json())
app.use(bodyParser.urlencoded({
  extended: true
}))

// Set up v6 app for backwards compatibility
if (useV6) {
  var v6Views = [
    path.join(__dirname, '/node_modules/govuk_template_jinja/views/layouts'),
    path.join(__dirname, '/app/v6/views/'),
    path.join(__dirname, '/lib/v6') // for old unbranded template
  ]
  nunjucksConfig.express = v6App
  var nunjucksV6Env = nunjucks.configure(v6Views, nunjucksConfig)

  // Nunjucks filters
  utils.addNunjucksFilters(nunjucksV6Env)

  // Set views engine
  v6App.set('view engine', 'html')

  // Backward compatibility with GOV.UK Elements
  app.use('/public/v6/', express.static(path.join(__dirname, '/node_modules/govuk_template_jinja/assets')))
  app.use('/public/v6/', express.static(path.join(__dirname, '/node_modules/govuk_frontend_toolkit')))
  app.use('/public/v6/javascripts/govuk/', express.static(path.join(__dirname, '/node_modules/govuk_frontend_toolkit/javascripts/govuk/')))
}

// Add global variable to determine if DoNotTrack is enabled.
// This indicates a user has explicitly opted-out of tracking.
// Therefore we can avoid injecting third-party scripts that do not respect this decision.
app.use(function (req, res, next) {
  // See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/DNT
  res.locals.doNotTrackEnabled = (req.header('DNT') === '1')
  next()
})

// Add variables that are available in all views
app.locals.gtmId = gtmId
app.locals.asset_path = '/public/'
app.locals.useAutoStoreData = (useAutoStoreData === 'true')
app.locals.useCookieSessionStore = (useCookieSessionStore === 'true')
app.locals.cookieText = config.cookieText
app.locals.promoMode = promoMode
app.locals.releaseVersion = 'v' + releaseVersion
app.locals.serviceName = config.serviceName
// extensionConfig sets up variables used to add the scripts and stylesheets to each page.
app.locals.extensionConfig = extensions.getAppConfig()

// Session uses service name to avoid clashes with other prototypes
const sessionName = 'govuk-prototype-kit-' + (Buffer.from(config.serviceName, 'utf8')).toString('hex')
let sessionOptions = {
  secret: sessionName,
  cookie: {
    maxAge: 1000 * 60 * 60 * 4, // 4 hours
    secure: isSecure
  }
}

// Support session data in cookie or memory
if (useCookieSessionStore === 'true') {
  app.use(sessionInCookie(Object.assign(sessionOptions, {
    cookieName: sessionName,
    proxy: true,
    requestKey: 'session'
  })))
} else {
  app.use(sessionInMemory(Object.assign(sessionOptions, {
    name: sessionName,
    resave: false,
    saveUninitialized: false
  })))
}

// Automatically store all data users enter
if (useAutoStoreData === 'true') {
  app.use(utils.autoStoreData)
  utils.addCheckedFunction(nunjucksAppEnv)
  if (useDocumentation) {
    utils.addCheckedFunction(nunjucksDocumentationEnv)
  }
  if (useV6) {
    utils.addCheckedFunction(nunjucksV6Env)
  }
}

// Clear all data in session if you open /prototype-admin/clear-data
app.post('/prototype-admin/clear-data', function (req, res) {
  req.session.data = {}
  res.render('prototype-admin/clear-data-success')
})

// Redirect root to /docs when in promo mode.
if (promoMode === 'true') {
  console.log('Prototype Kit running in promo mode')

  app.locals.cookieText = 'GOV.UK uses cookies to make the site simpler. <a href="/docs/cookies">Find out more about cookies</a>'

  app.get('/', function (req, res) {
    res.redirect('/docs')
  })

  // Allow search engines to index the Prototype Kit promo site
  app.get('/robots.txt', function (req, res) {
    res.type('text/plain')
    res.send('User-agent: *\nAllow: /')
  })
} else {
  // Prevent search indexing
  app.use(function (req, res, next) {
    // Setting headers stops pages being indexed even if indexed pages link to them.
    res.setHeader('X-Robots-Tag', 'noindex')
    next()
  })

  app.get('/robots.txt', function (req, res) {
    res.type('text/plain')
    res.send('User-agent: *\nDisallow: /')
  })
}

// Load routes (found in app/routes.js)
if (typeof (routes) !== 'function') {
  console.log(routes.bind)
  console.log('Warning: the use of bind in routes is deprecated - please check the Prototype Kit documentation for writing routes.')
  routes.bind(app)
} else {
  app.use('/', routes)
}

if (useDocumentation) {
  // Clone app locals to documentation app locals
  // Use Object.assign to ensure app.locals is cloned to prevent additions from
  // updating the original app.locals
  documentationApp.locals = Object.assign({}, app.locals)
  documentationApp.locals.serviceName = 'Prototype Kit'

  // Create separate router for docs
  app.use('/docs', documentationApp)

  // Docs under the /docs namespace
  documentationApp.use('/', documentationRoutes)
}

app.use(function (req, res, next) {
  nunjucksAppEnv.addGlobal('breadcrumbItems', function(text) {
    var items = [ { text: 'Becoming a teacher design history', href: '/' } ];

    if (req.path.match(/(teacher-training|manage-applications)\/?$/)) {
      return items;
    }

    if (req.path.includes('publish-teacher')) {
      items.push({
        text: 'Publish teacher training courses',
        href: '/publish-teacher-training'
      })
    }

    if (req.path.includes('find-teacher')) {
      items.push({
        text: 'Find postgraduate teacher training',
        href: '/find-teacher-training'
      })
    }

    if (req.path.includes('apply-teacher')) {
      items.push({
        text: 'Apply for postgraduate teacher training',
        href: '/apply-teacher-training'
      })
    }

    if (req.path.includes('manage-applications')) {
      items.push({
        text: 'Manage teacher training applications',
        href: '/manage-applications'
      })
    }

    return items;
  });

  nunjucksAppEnv.addGlobal('markdown', function(text) {
    if (text === undefined) {
      return ''
    }

    // Ignore leading indentation
    t = text.replace(/^\s+/gm, "\n")

    return '<div class="markdown">' + marked(t) + '</div>';
  });

  next()
});

if (useV6) {
  // Clone app locals to v6 app locals
  v6App.locals = Object.assign({}, app.locals)
  v6App.locals.asset_path = '/public/v6/'

  // Create separate router for v6
  app.use('/', v6App)

  // Docs under the /docs namespace
  v6App.use('/', v6Routes)
}

// Strip .html and .htm if provided
app.get(/\.html?$/i, function (req, res) {
  var path = req.path
  var parts = path.split('.')
  parts.pop()
  path = parts.join('.')
  res.redirect(path)
})

// Auto render any view that exists

// App folder routes get priority
app.get(/^([^.]+)$/, function (req, res, next) {
  utils.matchRoutes(req, res, next)
})

if (useDocumentation) {
  // Documentation  routes
  documentationApp.get(/^([^.]+)$/, function (req, res, next) {
    if (!utils.matchMdRoutes(req, res)) {
      utils.matchRoutes(req, res, next)
    }
  })
}

if (useV6) {
  // App folder routes get priority
  v6App.get(/^([^.]+)$/, function (req, res, next) {
    utils.matchRoutes(req, res, next)
  })
}

// Redirect all POSTs to GETs - this allows users to use POST for autoStoreData
app.post(/^\/([^.]+)$/, function (req, res) {
  res.redirect('/' + req.params[0])
})

// Catch 404 and forward to error handler
app.use(function (req, res, next) {
  var err = new Error(`Page not found: ${req.path}`)
  err.status = 404
  next(err)
})

// Display error
app.use(function (err, req, res, next) {
  console.error(err.message)
  res.status(err.status || 500)
  res.send(err.message)
})

console.log('\nGOV.UK Prototype Kit v' + releaseVersion)
console.log('\nNOTICE: the kit is for building prototypes, do not use it for production services.')

module.exports = app


app/config.js
// Use this file to change prototype configuration.

// Note: prototype config can be overridden using environment variables (eg on heroku)

module.exports = {
  // Service name used in header. Eg: 'Renew your passport'
  serviceName: 'Become a teacher design history',

  // Default port that prototype runs on
  port: '3000',

  // Enable or disable password protection on production
  useAuth: 'false',

  // Automatically stores form data, and send to all views
  useAutoStoreData: 'true',

  // Enable cookie-based session store (persists on restart)
  // Please note 4KB cookie limit per domain, cookies too large will silently be ignored
  useCookieSessionStore: 'false',

  // Enable or disable built-in docs and examples.
  useDocumentation: 'false',

  // Force HTTP to redirect to HTTPS on production
  useHttps: 'true',

  // Cookie warning - update link to service's cookie page.
  cookieText: 'GOV.UK uses cookies to make the site simpler. <a href="#">Find out more about cookies</a>',

  // Enable or disable Browser Sync
  useBrowserSync: 'true'

}


scripts/generate.js
/*
  Usage:
  * Put images into `app/assets/directory-name`
  * Name them with 01-, 02- prefixes

  Run:
  node scripts/generate.js name-of-directory-holding-images
*/

// Dependencies
const fs = require('fs');
const sharp = require('sharp');

// Arguments
const directoryName = process.argv.slice(-1)[0];
warnIfNoArguments(title);

var title = directoryName.replace(/-/g, ' ');
title = title.charAt(0).toUpperCase() + title.slice(1)
const imageDirectory = `app/assets/images/${directoryName}`;
const indexDirectory = `app/views/${directoryName}`;
const thumbnailDirectory = `${imageDirectory}/thumbnails`;

var paths = [];

// Run
function start() {
  makeDirectories();
  getExistingImages();
  generatePage();
  generateThumbnails();
}

function warnIfNoArguments(title) {
  // TODO: Use a better check for an argument
  if (directoryName.startsWith('/Users')) {
    console.log('No arguments set');
    console.log('Please set a title: `node scripts/screenshot.js "Title of page"`');
    return;
  }
}

function makeDirectories() {
  if (!fs.existsSync(imageDirectory)){
    fs.mkdirSync(imageDirectory);
  }

  if (!fs.existsSync(thumbnailDirectory)){
    fs.mkdirSync(thumbnailDirectory);
  }

  if (!fs.existsSync(indexDirectory)){
    fs.mkdirSync(indexDirectory);
  }
}

function getExistingImages() {
  const files = fs.readdirSync(imageDirectory);

  files.forEach((file, index) => {
    if (!(/^\d{2}/.test(file) && /\.(png|jpg)$/.test(file))) {
      console.log('Ignoring: ' + file);
      return;
    }

    var id = file.replace(/\.(png|jpg)$/, '');
    var title = id.replace(/^\d{2}-/, '').replace(/-/g, ' ');

    var image = {
      title: title.charAt(0).toUpperCase() + title.slice(1),
      id: id,
      file: `${imageDirectory}/${file}`,
      thumbnailFile: `${thumbnailDirectory}/${file}`
    }

    image.src = image.file.replace('app/assets', '/public');
    image.thumbnailSrc = image.thumbnailFile.replace('app/assets', '/public');

    paths.push(image);
  });
}

function generatePage() {
  var template = '';
  const templateStart = `{% extends "layout.html" %}
{% set title = '${title}' %}
{% block pageTitle %}{{ title }}{% endblock %}
{% block breadcrumbs %}{{ designHistory.breadcrumbs(breadcrumbItems()) }}{% endblock %}

{% block content %}
  <h1 class="govuk-heading-xl">{{ title }}</h1>
`;

  const templateEnd = `
{% endblock %}
`;

  var contents = `
  {% set contents = [`;

  const endContents = `
  ] %}
  {{ designHistory.screenshotContents(contents) }}
  `;

  paths.forEach(function(item, index) {
    template += `
  {{ designHistory.screenshot('${item.title}', '${item.id}', '${item.thumbnailSrc}', '${item.src}', '') }}
`;

    contents += `${index > 0 ? ', ': ''}
    { text: '${item.title}', id: '${item.id}' }`;
  });

  fs.writeFile(
    `${indexDirectory}/index.html`,
    templateStart + contents + endContents + template + templateEnd,
    function(err) {
      if (err) { return console.log(err); }
      console.log(`Index generated: ${indexDirectory}/index.html`);
    }
  );
}

function generateThumbnails() {
  paths.forEach(function(item, index) {
    sharp(item.file).resize(630, null).toFile(item.thumbnailFile);
  });
}

start();


scripts/screenshot.js
/*
  Usage:
  Set domain to the website you want to screenshot, eg localhost:3000
  Set paths to an array of named paths (see example)

  Run:
  node scripts/screenshot.js name-of-directory

  Example paths:
  paths = [
    { title: 'Index page', path: '/'},
    { title: 'Terms and conditions', path: '/terms-conditions'}
  ]
*/
const paths = [
  { title: 'Design history index', path: '/'}
];
const domain = 'https://design-history.herokuapp.com';

// Dependencies
const webshot = require('webshot');
const fs = require('fs');
const sharp = require('sharp');

// Arguments
const directoryName = process.argv.slice(-1)[0];
warnIfNoArguments();

var title = directoryName.replace(/-/g, ' ');
title = title.charAt(0).toUpperCase() + title.slice(1)
const imageDirectory = `app/assets/images/${directoryName}`;
const indexDirectory = `app/views/${directoryName}`;
const thumbnailDirectory = `${imageDirectory}/thumbnails`;

// Run
function start() {
  makeDirectories();
  decoratePaths();
  generatePage();
  takeScreenshots();
}

function warnIfNoArguments(title) {
  // TODO: Use a better check for an argument
  if (directoryName.startsWith('/Users')) {
    console.log('No arguments set');
    console.log('Please set a directory name: `node scripts/screenshot.js "name-of-directory"`');
    return;
  }
}

function makeDirectories() {
  if (!fs.existsSync(imageDirectory)){
    fs.mkdirSync(imageDirectory);
  }

  if (!fs.existsSync(thumbnailDirectory)){
    fs.mkdirSync(thumbnailDirectory);
  }

  if (!fs.existsSync(indexDirectory)){
    fs.mkdirSync(indexDirectory);
  }
}

function decoratePaths() {
  paths.forEach(function(item, index) {
    item.id = item.title.replace(/ +/g, '-').toLowerCase();
    item.file = `${imageDirectory}/${item.id}.png`;
    item.thumbnailFile = `${thumbnailDirectory}/${item.id}.png`;
    item.src = item.file.replace('app/assets', '/public');
    item.thumbnailSrc = item.thumbnailFile.replace('app/assets', '/public');
  });
}

function takeScreenshots() {
  // https://github.com/brenden/node-webshot
  const webshotOptions = {
    phantomConfig: {
      'ignore-ssl-errors': 'true'
    },
    windowSize: {
      width: 1200,
      height: 800
    },
    shotSize: {
      width: 'window',
      height: 'all'
    }
  }

  paths.forEach(function(item, index) {
    webshot(
      domain + item.path,
      item.file,
      webshotOptions,
      function(err) {
        sharp(item.file).resize(630, null).toFile(item.thumbnailFile);
        console.log(`${domain + item.path} \n >> ${item.file}`);
      }
    );
  });
}

function generatePage() {
  var template = '';
  const templateStart = `{% extends "layout.html" %}
{% set title = '${title}' %}
{% block pageTitle %}{{ title }}{% endblock %}
{% block breadcrumbs %}{{ designHistory.breadcrumbs(breadcrumbItems()) }}{% endblock %}

{% block content %}
  <h1 class="govuk-heading-xl">{{ title }}</h1>
`;

  const templateEnd = `
{% endblock %}
`;

  var contents = `
  {% set contents = [`;

  const endContents = `
  ] %}
  {{ designHistory.screenshotContents(contents) }}
  `;

  paths.forEach(function(item, index) {
    template += `
  {{ designHistory.screenshot('${item.title}', '${item.id}', '${item.thumbnailSrc}', '${item.src}', '') }}
`;

    contents += `${index > 0 ? ', ': ''}
    { text: '${item.title}', id: '${item.id}' }`;
  });

  fs.writeFile(
    `${indexDirectory}/index.html`,
    templateStart + contents + endContents + template + templateEnd,
    function(err) {
      if (err) { return console.log(err); }
      console.log(`Index generated: ${indexDirectory}/index.html`);
    }
  );
}

start();
